### ISSUE TITLE
Handling Extreme Class Imbalance in Custom YOLOv8 Dataset

### ISSUE THREAD

**Issue by NEREUScode** [ultralytics/ultralytics#20259]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone! I'm training a custom object detection model using YOLOv8 on a dataset that contains a significant class imbalance, and I wanted to get some advice on the best way to handle it.

Here's a breakdown of the class distribution in my dataset:

- **Bottle**: 719  
- **Plastic fragment**: 356  
- **Cap**: 195  
- **Metal**: 6  
- **Cup**: 10  
- **Glass bottle**: 79  
- **Can**: 23  
- **Tire**: 19  
- **Wood**: 101  
- **Tin can**: 12  
- **Net**: 68  
- **Bag**: 26  
- **Paper**: 2  
- **Cardboard**: 4  
- **Straw**: 1  

As you can see, some classes like **Bottle** have hundreds of examples, while others like **Straw**, **Paper**, and **Cardboard** have fewer than 5. I'm worried that this imbalance will cause YOLO to train poorly on the underrepresented classes or even ignore them entirely during inference.

One of the strategies I'm currently considering is to **apply targeted augmentation only on images containing rare classes**, to try and bring the dataset closer to a balance without overfitting the dominant classes. However, I'm unsure how effective this will be, especially for extremely rare cases like **Straw** (just one instance). I can't find more real samples for these rare classes at the moment, so I'm not sure what the best alternative is.

---

### ❓ My Questions:
- Is augmenting only the images that contain minority classes a good approach?
- For classes with **extremely few samples** (like 1 or 2), is there any YOLO-specific trick or technique that helps in such cases (e.g., synthetic generation, class weighting, etc.)?
- Would it make sense to **temporarily exclude these ultra-rare classes** to improve model focus, or will YOLO still learn something useful from a single example?
- Is it recommended to use class weights or focal loss with YOLOv8 in this context?

Any thoughts, experience, or suggestions would be super helpful. Thanks a lot in advance!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @NEREUScode, thank you for reaching out and sharing your detailed dataset breakdown with us! 🚀 This is an automated response to help you get started, and an Ultralytics engineer will review and assist you further soon.

We recommend starting with the [Docs](https://docs.ultralytics.com/) where you can find guides on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, including tips on data preparation and augmentation strategies. Many common questions around class imbalance and training best practices are addressed there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further. For custom training ❓ Questions like yours, sharing additional dataset examples or training logs can help us deliver more targeted advice. Make sure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for discussion and real-time support:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for chat
- [Discourse](https://community.ultralytics.com/) for in-depth topics
- [Subreddit](https://reddit.com/r/Ultralytics) to connect with other users

## Upgrade

Before proceeding, ensure you are running the latest release of `ultralytics` and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your question and for using Ultralytics! An engineer will follow up to provide additional guidance on handling extreme class imbalance and your specific training questions. 😊

**Comment by Y-T-G** [comment]
You can try this
https://y-t-g.github.io/tutorials/yolo-class-balancing/

But the lower number of samples is a bigger issue here than imbalance. Artificially increasing the sample through augmentations wouldn't help with the lack of diversity which would affect the model's ability to generalize. If anything, it would cause overfitting.

**Comment by Jordan-Pierce** [comment]
@NEREUScode can confirm that @Y-T-G 's method works well (but it's also not a miracle worker)

**Comment by Jordan-Pierce** [comment]
@Y-T-G, have you considered making a variation of the `yolo-class-balancing` class that focuses on samples in the training set that are more prone to errors / larger losses? Instead of just increasing the (probability) of the frequency minority classes are shown, include those that have the largest loss in an epoch?

**Comment by Y-T-G** [comment]
@Jordan-Pierce I think that would require deeper modifications in several places to make it work. It's also complicated by mosaic augmentation which mixes 4 images into one.

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082
